sim/state.ts

```diff
@@ -0,0 +1,340 @@
+		const battle = new Battle(options);
+		this.deserialize(state, battle, BATTLE, battle);
+		this.deserializeField(state.field, battle.field);
+		for (const [i, side] of state.sides.entries()) {
+			this.deserializeSide(side, battle.sides[i]);
+		}
+		battle.prng = new PRNG(state.prng);
+		// @ts-ignore - readonly
+		battle.hints = new Set(battle.hints);
+		return battle;
+	}
+
+	serializeField(field: Field): /* Field */ AnyObject {
+		return this.serialize(field, FIELD, field.battle);
+	}
+
+	private deserializeField(state: /* Field */ AnyObject, field: Field) {
+		this.deserialize(state, field, FIELD, field.battle);
+	}
+
+	serializeSide(side: Side): /* Side */ AnyObject {
+		const state: /* Side */ AnyObject = this.serialize(side, SIDE, side.battle);
+		state.pokemon = new Array(side.pokemon.length);
+		const team = new Array(side.pokemon.length);
+		for (const [i, pokemon] of side.pokemon.entries()) {
+			state.pokemon[i] = this.serializePokemon(pokemon);
+			team[side.team.indexOf(pokemon.set)] = i + 1;
+		}
+		state.team = team.join(team.length > 9 ? ',' : '');
+		state.choice = this.serializeChoice(side.choice, side.battle);
+		return state;
+	}
+
+	private deserializeSide(state: /* Side */ AnyObject, side: Side) {
+		this.deserialize(state, side, SIDE, side.battle);
+		const ordered = new Array(side.pokemon.length);
+		const team = state.team.split(state.team.length > 9 ? ',' : '');
+		for (const [i, pos] of team.entries()) {
+			ordered[Number(pos) - 1] = side.pokemon[i];
+		}
+		side.pokemon = ordered;
+		for (const [i, pokemon] of state.pokemon.entries()) {
+			this.deserializePokemon(pokemon, side.pokemon[i]);
+		}
+		this.deserializeChoice(state.choice, side.choice, side.battle);
+	}
+
+	serializePokemon(pokemon: Pokemon): /* Pokemon */ AnyObject {
+		const state: /* Pokemon */ AnyObject = this.serialize(pokemon, POKEMON, pokemon.battle);
+		state.set = pokemon.set;
+		return state;
+	}
+
+	private deserializePokemon(state: /* Pokemon */ AnyObject, pokemon: Pokemon) {
+		this.deserialize(state, pokemon, POKEMON, pokemon.battle);
+		// @ts-ignore - readonly
+		pokemon.set = state.set;
+	}
+
+	private serializeChoice(choice: Choice, battle: Battle): /* Choice */ AnyObject {
+		const state: /* Choice */ AnyObject = this.serialize(choice, CHOICE, battle);
+		state.switchIns = Array.from(choice.switchIns);
+		return state;
+	}
+
+	private deserializeChoice(state: /* Choice */ AnyObject, choice: Choice, battle: Battle) {
+		this.deserialize(state, choice, CHOICE, battle);
+		choice.switchIns = new Set(state.switchIns);
+	}
+
+	// Simply looking for a 'hit' field to determine if an object is an ActiveMove or not seems
+	// pretty fragile, but its no different than what the simulator is doing. We go further and
+	// also check if the object has an 'id', as that's what we will intrepret as the Move.
+	private isActiveMove(obj: AnyObject): obj is ActiveMove {
+		return obj.hasOwnProperty('hit') && obj.hasOwnProperty('id');
```

Comment:
Why don't you check for ``isCopy``, which is *exactly* what the simulator is doing?

Reply:
Um, what?

https://github.com/Zarel/Pokemon-Showdown/blob/master/sim/dex.ts#L458

Reply:
Uh? Fine then, I wonder why would @Zarel change this...